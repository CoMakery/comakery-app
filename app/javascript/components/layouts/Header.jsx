import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Icon from './../styleguide/Icon'

const navContent = (isLoggedIn, isAdmin, isWhitelabel, projectAwardsVisible, currentPath) => {
  return <React.Fragment>
    { !isLoggedIn &&
      <div className="header--nav--links">
        { !isWhitelabel &&
          <a href="/projects" className={(currentPath === '/projects') ? 'header--nav--links--current' : null}>
            Projects
          </a>
        }

        { !isWhitelabel &&
          <a href="https://forum.comakery.com" target="_blank">
            Forum
          </a>
        }

        { !isWhitelabel &&
          <a href="https://ledger.comakery.com" target="_blank">
            Blog
          </a>
        }

        { !isWhitelabel &&
          <a target="_blank" href="https://info.comakery.com/pricing/">
            Pricing
          </a>
        }

        { !isWhitelabel &&
          <a target="_blank" href="https://info.comakery.com/about_us/">
            About Us
          </a>
        }

        <a href="/session/new" className={currentPath.match(/session\/new/) ? 'header--nav--links--current header--nav--links--sign' : null} >
          Sign In
        </a>

        <a href="/accounts/new" className={currentPath.match(/accounts\/new/) ? 'header--nav--links--current header--nav--links--sign' : null} >
          Sign Up
        </a>
      </div>
    }

    { isLoggedIn &&
      <div className="header--nav--links">
        { isAdmin &&
          <React.Fragment>
            <a href="/accounts" className={currentPath.match(/^\/accounts$/) ? 'header--nav--links--current' : null}>
              All Accounts
            </a>
            <a href="/missions" className={currentPath.match(/missions/) ? 'header--nav--links--current' : null}>
              Missions Admin
            </a>

            <a href="/tokens" className={currentPath.match(/tokens/) ? 'header--nav--links--current' : null}>
              Tokens Admin
            </a>
          </React.Fragment>
        }

        <a href="/projects/mine" className={currentPath.match(/(?!.*tasks.*)projects.*/) ? 'header--nav--links--current' : null} >
          My Projects
        </a>

        { (!isWhitelabel || projectAwardsVisible) &&
          <a href="/tasks" className={currentPath.match(/\/tasks/) ? 'header--nav--links--current' : null}>
            My Tasks
          </a>
        }

        { !isWhitelabel &&
          <a href="https://forum.comakery.com" target="_blank">
            Forum
          </a>
        }

        { !isWhitelabel &&
          <a href="https://ledger.comakery.com" target="_blank">
            Blog
          </a>
        }

        <a href="/wallets" className={currentPath.match(/\/(wallets)/) ? 'header--nav--links--current' : null} >
          Wallets
        </a>

        <a href="/account" className={currentPath.match(/\/(account$)/) ? 'header--nav--links--current' : null} >
          My Account
        </a>

        <a rel="nofollow" href="/session" data-method="delete" data-turbo="false">
          Sign out
        </a>
      </div>
    }
  </React.Fragment>
}

class Header extends React.Component {
  constructor(props) {
    super(props)
    this.state = { mobileMenuActive: false }
  }
  render() {
    const {className, isAdmin, isLoggedIn, isWhitelabel, whitelabelLogo, projectAwardsVisible, currentPath, ...other} = this.props

    const classnames = classNames(
      'header',
      className
    )

    const {mobileMenuActive} = this.state
    const nav = navContent(isLoggedIn, isAdmin, isWhitelabel, projectAwardsVisible, currentPath)

    return (
      <React.Fragment>
        <div className={classnames} {...other}>
          <div className="header--logo">
            <a href="/">
              { !isWhitelabel &&
                <Icon name="Logo-Header.svg" width="200px" />
              }

              { isWhitelabel &&
                <img src={whitelabelLogo} width="200px" />
              }
            </a>
          </div>

          <div className="header--nav">
            {nav}
          </div>
        </div>

        <div className="header-mobile">
          { !isWhitelabel &&
            <Icon name="Logo-Header.svg" width="162px" />
          }

          { isWhitelabel &&
            <img src={whitelabelLogo} width="162px" />
          }

          <div className="header-mobile__menu-icon" onClick={() => {
            this.setState({mobileMenuActive: !mobileMenuActive})
            document.body.classList.toggle('body--overflow')
          }} >
            <span className={mobileMenuActive ? 'active' : ''} />
          </div>

          { mobileMenuActive &&
            <div className="header-mobile__menu">
              {nav}
            </div>
          }
        </div>
      </React.Fragment>
    )
  }
}
Header.propTypes = {
  currentPath         : PropTypes.string,
  isAdmin             : PropTypes.bool,
  isLoggedIn          : PropTypes.bool,
  projectAwardsVisible: PropTypes.bool,
}
Header.defaultProps = {
  currentPath         : '/',
  isAdmin             : false,
  isLoggedIn          : false,
  projectAwardsVisible: false,
}
export default Header
