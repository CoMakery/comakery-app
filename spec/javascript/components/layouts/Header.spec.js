import React from 'react'
import { shallow } from 'enzyme'
import Header from 'components/layouts/Header'

describe('Header', () => {
  it('renders correctly without props', () => {
    const wrapper = shallow(<Header isLoggedIn currentPath="/projects/mine" />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.header')).toBe(true)
    expect(wrapper.exists('.header--logo')).toBe(true)
    expect(wrapper.exists('.header--nav')).toBe(true)
    expect(wrapper.exists('.header--nav--links')).toBe(true)
    expect(wrapper.find('.header--nav--links').text()).not.toMatch(/Missions Admin/)
    expect(wrapper.find('.header--nav--links').text()).not.toMatch(/Tokens Admin/)
  })

  it('renders correctly with signup path and without loggedIn flag', () => {
    const wrapper = shallow(<Header currentPath="/accounts/new" />)

    expect(wrapper.find('.header--nav--links--current').text()).toBe('Sign Up')
    expect(wrapper.find('.header--nav--links--current').props().href).toBe('/accounts/new')
  })

  it('renders correctly with signin path and without loggedIn flag', () => {
    const wrapper = shallow(<Header currentPath="/session/new" />)

    expect(wrapper.find('.header--nav--links--current').text()).toBe('Sign In')
    expect(wrapper.find('.header--nav--links--current').props().href).toBe('/session/new')
  })

  it('renders correctly with root path and loggedIn flag', () => {
    const wrapper = shallow(<Header isLoggedIn currentPath="/projects/mine" />)

    expect(wrapper.find('.header--nav--links--current').props().href).toBe('/projects/mine')
  })

  it('renders correctly with My Account path and loggedIn flag', () => {
    const wrapper = shallow(<Header isLoggedIn currentPath="/account" />)

    expect(wrapper.find('.header--nav--links--current').text()).toBe('My Account')
    expect(wrapper.find('.header--nav--links--current').props().href).toBe('/account')
  })

  it('renders correctly with My Projects path and loggedIn flag', () => {
    const wrapper = shallow(<Header isLoggedIn currentPath="/projects/mine" />)

    expect(wrapper.find('.header--nav--links--current').text()).toBe('My Projects')
    expect(wrapper.find('.header--nav--links--current').props().href).toBe('/projects/mine')
  })

  it('renders correctly with admin flag', () => {
    const wrapper = shallow(<Header isLoggedIn isAdmin />)

    expect(wrapper.find('.header--nav--links').text()).toMatch(/Missions Admin/)
    expect(wrapper.find('.header--nav--links').text()).toMatch(/Tokens Admin/)
  })

  it('renders correctly with custom className', () => {
    const wrapper = shallow(<Header className="__test" />)

    expect(wrapper.exists('.header.__test')).toBe(true)
  })
})
