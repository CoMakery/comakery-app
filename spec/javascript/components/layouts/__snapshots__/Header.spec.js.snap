// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly without props 1`] = `
<Fragment>
  <div
    className="header"
  >
    <div
      className="header--logo"
    >
      <a
        href="/"
      >
        <Icon
          className=""
          name="Logo-Header.svg"
          width="200px"
        />
      </a>
    </div>
    <div
      className="header--nav"
    >
      <div
        className="header--nav--links"
      >
        <a
          className="header--nav--links--current"
          href="/projects/mine"
        >
          My Projects
        </a>
        <a
          className={null}
          href="/tasks"
        >
          My Tasks
        </a>
        <a
          href="https://forum.comakery.com"
          target="_blank"
        >
          Forum
        </a>
        <a
          href="https://ledger.comakery.com"
          target="_blank"
        >
          Blog
        </a>
        <a
          className={null}
          href="/wallets"
        >
          Wallets
        </a>
        <a
          className={null}
          href="/account"
        >
          My Account
        </a>
        <a
          data-method="delete"
          data-turbo="false"
          href="/session"
          rel="nofollow"
        >
          Sign out
        </a>
      </div>
    </div>
  </div>
  <div
    className="header-mobile"
  >
    <Icon
      className=""
      name="Logo-Header.svg"
      width="162px"
    />
    <div
      className="header-mobile__menu-icon"
      onClick={[Function]}
    >
      <span
        className=""
      />
    </div>
  </div>
</Fragment>
`;
